### DISCUSSION TITLE
[v4] Tailwind does not parse Svelte class directives

### DISCUSSION THREAD

**Discussion by omaishar** [tailwindlabs/tailwindcss#15312] [Help]
Tailwind v4 does not seem to parse any [Svelte class directives](https://svelte.dev/tutorial/svelte/classes).
Example:

Tailwind v3 - works!
`<div class:bg-gray-500={isDisabled} ...>`
After running `tailwindcss` CLI command the `bg-gray-500` class is added to the `app.min.css` file.

Tailwind v4 - does not work!
`<div class:bg-gray-500={isDisabled} ...>`
After running `npx @tailwindcss/cli` CLI command the `bg-gray-500` class is NOT added to the `app.min.css` file.

**What version of Tailwind CSS are you using?**
Tailwind: 4.0.0-beta.4
Tailwind CLI: 4.0.0-beta.4

**What build tool (or framework if it abstracts the build tool) are you using?**
SvelteKit: 2.5.26
Vite: 5.4.4

**What version of Node.js are you using?**
20.17.0

**What browser are you using?**

For example: Chrome, Safari, or N/A

**What operating system are you using?**
MacOS

**Comment by philipp-spiess** [comment]
Hey @omaishar! Yeah unfortunately slashes inside class names are not supported by the class directive in Svelte right now. 

You can make it work around this limitation by adding classes like this, though:

```svelte
<div class="{ isDisabled ? 'bg-gray-500/50' : '' }"></div>
```

I don't think there's going to be anything we can do to fix this from our end but there seems to be appetite like there's interest in allowing more characters in css class names if you take a look at this issue: https://github.com/sveltejs/svelte/issues/7170

**Reply by omaishar** [reply]
@philipp-spiess Your answer is not relevant to this discussion (which should really be an issue, not a discussion) but rather to [[v4] New color opacity syntax not working with Svelte class directives](https://github.com/tailwindlabs/tailwindcss/discussions/15310).

The issue here is not related to the opacity, but to the fact that Tailwind v4 no longer parses Svelte class directives at all, as opposed to Tailwind v3 which does. This basically breaks almost any codebase based on Svelte and Tailwind.

**Reply by philipp-spiess** [reply]
@omaishar Ah gotcha you're right that answer was indeed pointed to the other discussion, I didn't realize you opened two. Let me look into this one then. Sorry!

**Reply by philipp-spiess** [reply]
Do you mind sharing a repro of this? I've tried to recreate this with the latest version of the CLI and the class name is picked up as expected:

<img width="1745" alt="Screenshot 2024-12-06 at 12 57 41" src="https://github.com/user-attachments/assets/9b36c4d0-a7c2-447c-83a6-8934528756cd">

**Reply by notramo** [reply]
@philipp-spiess, could you try it in the Vite dev server? I also experience the issue with latest SvelteKit, Vite and Tailwind v4 beta.

**Reply by philipp-spiess** [reply]
@notramo I did try it with both a direct Vite devserver and Sveltekit (also using Vite). A reproduction setup would help if you can set something up. Some folks have reported issues with the order of how they register Vite extension, so one thing you can try to do ist to reorder the Svelte/Tailwind CSS extensions?

**Reply by notramo** [reply]
@philipp-spiess, thanks, switching the plugin order solved it. The SvelteKit plugin must be first, followed by Tailwind. When Tailwind is the first, it fails to detect. Does the docs mention this somewhere?

works:
```typescript
export default defineConfig({
  plugins: [enhancedImages(), sveltekit(), tailwindcss()],
});
```

fails, this was my original config:
```typescript
export default defineConfig({
  plugins: [tailwindcss(), enhancedImages(), sveltekit()],
});
```

**Reply by omaishar** [reply]
There is still an issue here. It seems that Tailwind CLI successfully parse class directives now, but not for `<img>` elements.

Tailwind v3
`<div class:bg-gray-500={isDisabled} ...>`       - works!
`<img class:bg-gray-600={isDisabled} ...>`      - works!
After running `tailwindcss` CLI command the `bg-gray-500` and `bg-gray-600` class are both added to the app.min.css file.

Tailwind v4
`<div class:bg-gray-500={isDisabled} ...>`       - works!
`<img class:bg-gray-600={isDisabled} ...>`       - does NOT work!
After running `npx @tailwindcss/cli` CLI command the `bg-gray-500` class is added to the app.min.css file, but the class `bg-gray-600` is not.

It may also fail for other elements as well. I haven't checked all.

**Reply by philipp-spiess** [reply]
@notramo Ahh, thanks, yeah we'll get this cleared up in the docs.

@omaishar Can you share a repro for this? I've tested it with the following minimal repro and it does create both classes:

```css
/* input.css */
@theme {
   --color-gray-500: oklch(0.551 0.027 264.364);
   --color-gray-600: oklch(0.446 0.03 256.802);
}
@tailwind utilities;
```

```svelte
/* input.svelte */
<div class:bg-gray-500={isDisabled}>
  <img class:bg-gray-600={isDisabled} />
</div>
```

```css
/* out.css */
/*! tailwindcss v4.0.0-beta.6 | MIT License | https://tailwindcss.com */
:root {
  --color-gray-500: oklch(0.551 0.027 264.364);
  --color-gray-600: oklch(0.446 0.03 256.802);
}
.bg-gray-500 {
  background-color: var(--color-gray-500);
}
.bg-gray-600 {
  background-color: var(--color-gray-600);
}
```

Using the following command:

```bash
npx @tailwindcss/cli@next -i input.css -o out.css
```

My only guess is that the `<img>` tag in your example is in a file that's not scanned perhaps?

**Reply by omaishar** [reply]
Unfortunately our repos are all private.
The `<img>` tag is in a file that is being scanned. In fact, if I replace the `<img>` tag with a `<div>` it works fine.

**Reply by philipp-spiess** [reply]
Are you able to reproduce this in a fresh Svelte setup somehow? Not really sure how to investigate this issue otherwise. Happy to take the file that contains the `<img>` tag that won't have it's classes generated privately as well if you're able to share just that.

**Comment by christopher356** [comment]
Tailwind CSS v4 requires a different approach to class directives in Svelte. Use the class directive with dynamic class lists to achieve the desired behavior. For example: <div class="{isDisabled ? 'bg-gray-500' : ''}"></div>. This approach ensures that Tailwind CSS v4 correctly processes the class directives.
